### PR TITLE
Add math special functions: erf, erfcx, expint1, Bessel functions, Hankel functions

### PR DESCRIPTION
**Pull request by vqd8a**
This PR will add mathematical special functions needed by GEMMA's deep slot code.

### CODE REVIEW

**Comment by vqd8a** [comment]
I have added the implementations of these functions and their unit tests. All checks passed except HIP tests in jenkin-ci check, where all 0s are returned for the Bessel functions and Hankel functions. 
I don't know what happened since these unit tests are going well on Caraway and Tulip.

**Comment by Rombur** [comment]
Which compiler are you using on caraway and tulip? There are know bugs in the compiler used for CI. Once we merged https://github.com/kokkos/kokkos/pull/4025, the CI should used a better compiler.

**Comment by vqd8a** [comment]
@Rombur rocm/4.1.0 was used on caraway and rocm/4.2.0 was used on tulip. Thanks for letting me know. I will wait for #4025 got merged.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (dalg24) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:281]: Ok to add `erfcx` but we need to provide an implementation for scalar types as well
`erfcx(FloatingPoint)`

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (dalg24) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:457]: Implement `cyl_bessel_j` and error out if the order argument is not 0 (or 1)

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (dalg24) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:738]: Rename `cyl_bessel_y1` for consistency with the C++ math functions

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (vqd8a) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:65]: @dalg24 I see in Kokkos_Complex.hpp, taking argument by reference is used. Please explain why we should take argument by value.

Comment (dalg24) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:65]: Because this overload is for floating point numbers.
https://en.cppreference.com/w/cpp/numeric/special_functions/expint

Comment (vqd8a) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:65]: @dalg24 I see. Thanks.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (vqd8a) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:80]: @dalg24 I use the definition of E1(x) while std::expint uses the definition of Ei(x). So I changed the function name to ```expint1``` to avoid confusion.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (vqd8a) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp]: @dalg24 I have changed it to taking argument by value.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (vqd8a) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp]: @dalg24 The names are now ```erf``` and ```erfcx```.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (vqd8a) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp]: @dalg24 I have added an implementation for scalar type. For now, I do not use the std::erfc() for ```erfcx()=exp(x^2)*erfc(x)``` because it is going to overflow quickly (e.g. erfcx(35)=NaN). Instead, I use ```erfcx(Kokkos::complex<RealType>(x, 0.0))``` for now. I will have an implementation of ```erfcx(real)``` in the future.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (vqd8a) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp]: @dalg24 Since we decided  to use ```cyl_bessel_j0``` and ```cyl_bessel_j1``` for order 0 and order 1, respectively, so I do not have ```cyl_bessel_j``` for now.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp
Comment (vqd8a) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp]: @dalg24 I have changed all the Bessel function names accordingly.

**Comment by vqd8a** [comment]
@dalg24 I have addressed all your comments. Please let me know if you have additional comments. Thanks.

**Review by crtrott — APPROVED** [review]
I'd love for Damien to approve but not sure he got access right now.

**Comment by vqd8a** [comment]
@crtrott Thanks, Christian. Should we wait for Damien to merge it?

**Comment by crtrott** [comment]
I will check if he is able to review if not i will go ahead

**Pull request merged**
